### PR TITLE
feat: support being able to disable buttons

### PR DESCRIPTION
**Pull request by gkatsev**
The final piece I believe is to turn off thumbnails of the time-range when it's disabled. But wanted to get some 👀 to make sure that this direction is fine before making that change.

Another thing that I haven't been able to figure out is if I set up the following css:
```css
[disabled] {
  cursor: not-allowed;
}
```
When hovering over the `input` element itself, it would still show the hand cursor. Though, it will not do anything. I tried even adding a part on it and added `[disabled]::part(range)` to the selector above, and it still shows the hand cursor.

### CODE REVIEW

File: src/js/media-chrome-button.js [hunk 1]
```diff
@@ -91,30 +91,44 @@ class MediaChromeButton extends window.HTMLElement {
...
+  enable() {
+    this.addEventListener('click', this.#clickHandler);
+    this.addEventListener('keydown', this.#keydownHandler);
+    this.setAttribute('tabindex', 0);
+  }
+
+  disable() {
+    this.removeEventListener('click', this.#clickHandler);
+    this.removeEventListener('keyup', this.#keyUpHandler);
+    this.removeAttribute('tabindex');
```
Comment (gkatsev) [inline src/js/media-chrome-button.js:131]: Removing `tabindex` makes our buttons not focusable when they are disabled.

File: src/js/media-chrome-range.js [hunk 2]
```diff
@@ -213,6 +216,16 @@ class MediaChromeRange extends window.HTMLElement {
...
       }
+    } else if (
+      attrName === 'disabled' ||
+      attrName === 'aria-disabled' &&
+      oldValue !== newValue
+    ) {
+      if (newValue == null) {
+        this.range.removeAttribute(attrName);
+      } else {
+        this.range.setAttribute(attrName, newValue);
```
Comment (gkatsev) [inline src/js/media-chrome-range.js:227]: we want to copy over the `disabled` attributes onto the `input` to make sure that it gets disabled correctly.

**Review by luwes — APPROVED** [review]
👍

File: src/js/media-time-range.js [hunk 3]
```diff
@@ -313,77 +327,84 @@ class MediaTimeRange extends MediaChromeRange {
...
+    );
+    if (!this.#rangeEntered && mediaDurationStr) {
+      this.#rangeEntered = true;
+      this.#trackMouse();
+
+      window.addEventListener('pointermove', this.#offRangeHandler, false);
+    }
+  }
+
+  _enableBoxes() {
```
Comment (luwes) [inline src/js/media-time-range.js:399]: feel free to change these to private class methods too.
at the time I didn't know we were adopting these in Media Chrome.

File: src/js/media-time-range.js [hunk 4]
```diff
@@ -313,77 +327,84 @@ class MediaTimeRange extends MediaChromeRange {
...
+      !evt.composedPath().includes(this) ||
+      [...this.#boxes].some((b) => evt.composedPath().includes(b))
+    ) {
+      window.removeEventListener('pointermove', this.#offRangeHandler);
+      this.#rangeEntered = false;
+      this.#stopTrackingMouse();
+    }
+  }
+
+  #trackMouse = () => {
```
Comment (luwes) [inline src/js/media-time-range.js:374]: is this worth a separate method?

**Review by luwes — APPROVED** [review]
LGTM tested locally

File: src/js/media-time-range.js (hunk 3 above)
Comment (gkatsev) [reply src/js/media-time-range.js:399]: I guess worth input from others. I'm ok with either, since esbuild does compile them away.

Comment (gkatsev) [reply src/js/media-time-range.js:399]: The other private props are more important to have the event handler methods be per instance to avoid wardresses around removing the event handler and all buttons stopping working.

Comment (luwes) [reply src/js/media-time-range.js:399]: it's fine I think, support is pretty good https://caniuse.com/mdn-javascript_classes_private_class_fields

Comment (gkatsev) [reply src/js/media-time-range.js:399]: yeah, we could consider having esbuild not copmile it away.

File: src/js/media-chrome-button.js [hunk 5]
```diff
@@ -91,30 +91,44 @@ class MediaChromeButton extends window.HTMLElement {
...
-        this.removeEventListener('keyup', keyUpHandler);
-        return;
-      }
+    this.enable();
+  }
 
-      this.handleClick(e);
-    };
+  #clickHandler = (e) => {
+    this.handleClick(e);
```
Comment (heff) [inline src/js/media-chrome-button.js:98]: Two functions named basically the same thing isn't great. Maybe the addition of it being private is enough.

Now I'm wondering if it'd be better to have the child class set the click handler via `super()` instead of `.handleClick`.

File: src/js/media-chrome-button.js [hunk 6]
```diff
@@ -91,30 +91,44 @@ class MediaChromeButton extends window.HTMLElement {
...
-
-    // NOTE: There are definitely some "false positive" cases with multi-key pressing,
-    // but this should be good enough for most use cases.
-    const keyUpHandler = (e) => {
-      const { key } = e;
-      if (!this.keysUsed.includes(key)) {
-        this.removeEventListener('keyup', keyUpHandler);
-        return;
-      }
+    this.enable();
```
Comment (heff) [inline src/js/media-chrome-button.js:94]: Just confirming, we don't need to check if disabled already here, yeah?

File: src/js/media-chrome-range.js [hunk 7]
```diff
@@ -190,7 +190,10 @@ template.innerHTML = `
 
 class MediaChromeRange extends window.HTMLElement {
   static get observedAttributes() {
-    return [MediaUIAttributes.MEDIA_CONTROLLER];
+    return [
+      'disabled',
+      'aria-disabled',
```
Comment (heff) [inline src/js/media-chrome-range.js:195]: This is needed for `range` but not `button`?

File: src/js/media-time-range.js [hunk 8]
```diff
@@ -313,77 +327,84 @@ class MediaTimeRange extends MediaChromeRange {
     style.transform = `translateX(${boxPos}px)`;
   }
 
-  _enableBoxes() {
```
Comment (heff) [inline src/js/media-time-range.js:316]: It's hard to follow the diff after this point. Was this a simple moving around of code or was there a refactor happening?

**Review by heff — APPROVED** [review]
Looks good to me. Did just using `pointer-events: none` turn out to not be a solution here?

When finishing this up we should add the disabled state to the example files.

**Comment by gkatsev** [comment]
> Did just using `pointer-events: none` turn out to not be a solution here?

Yeah, `pointer-events: none` isn't enough here. First, we actually do want pointer interactions, at least partially. Since, we generally want to have `cursor: not-allowed` on disabled controls. Second, even if we set it and pointer events aren't allowed, keyboard and screen reader users would still be able to focus the element and interact with it which we don't want.

> When finishing this up we should add the disabled state to the example files.

Good call, I'll add something for that.

File: src/js/media-time-range.js (hunk 8 above)
Comment (gkatsev) [reply src/js/media-time-range.js:316]: yes, this mostly moves the handlers out of `_enableBoxes()` into private properties so that a `_disableBoxes()` method could be created.

File: src/js/media-chrome-button.js (hunk 5 above)
Comment (gkatsev) [reply src/js/media-chrome-button.js:98]: renamed to `#clickListener` along with `keyupListener` and `keydownListener`, for consistency. Hopefully, makes it clear enough that there's a difference.

**Review by heff — APPROVED** [review]

**Review by cjpillsbury — APPROVED** [review]
I think this is a reasonable initial implementation. Let's make sure we follow up on the Screen Reader/A11Y concerns/considerations/discussion we had out of band so it doesn't fall off completely. (tl;dr - Screen Reader-specific navigation vs. standard keyboard navigation and expectations/good experiences for folks relying on those)
